### PR TITLE
Display user global rank for multiplayer room participants

### PR DESCRIPTION
**Pull request by frenzibyte**
- [x] Depends on https://github.com/ppy/osu/pull/11799

image

(first user in above image has not submitted a score, hence no rank displayed; maybe display as `-`?)

For reason why this wasn't appearing, it's because it was relying on `CurrentModeRank`, which isn't passed in condensed user responses coming from `GetUsersRequest` (retrieved by `UserLookupCache`), therefore I've changed it to use the newly added `statistics_{osu/taiko/fruits/mania}` statistics instead.

The added dictionary property for retrieving `UserStatistics` from given ruleset should not exist in `User`, should be in a `UserCompact` class instead, but due to changes appear to be too complicated I've added a todo comment at the very least.

Visual test for this already exists, see `TestSceneMultiplayerParticipantsList.TestManyUsers()`.

### CODE REVIEW

File: osu.Game/Screens/OnlinePlay/Multiplayer/Participants/ParticipantPanel.cs [hunk 1]
```diff
@@ -162,6 +162,15 @@ protected override void OnRoomUpdated()
...
             const double fade_time = 50;
 
+            var ruleset = rulesets.GetRuleset(Room.Settings.RulesetID);
+
+            var currentModeRank = User.User?.GetStatisticsFor(ruleset)?.GlobalRank;
+
+            // fallback to current mode rank for testing purposes.
+            currentModeRank ??= User.User?.CurrentModeRank;
+
+            userRankText.Text = currentModeRank != null ? $"#{currentModeRank.Value:N0}" : string.Empty;
```
Comment (frenzibyte) [inline osu.Game/Screens/OnlinePlay/Multiplayer/Participants/ParticipantPanel.cs:172]: Not sure about this change (addition of `N0` at the end), previously it wasn't using commas in-between digits, but felt somewhat wrong, here's a *real-life* comparsion for reference:

Without commas:
image

With commas:
image

Comment (bdach) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Participants/ParticipantPanel.cs:172]: fyi `FormatUtils.FormatRank()` exists.

Comment (frenzibyte) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Participants/ParticipantPanel.cs:172]: That method is used for score ranks, not user global ranks. Don't think it's fine displaying something like `1k` or `1.5m`.

File: osu.Game/Screens/OnlinePlay/Multiplayer/Participants/ParticipantPanel.cs [hunk 2]
```diff
@@ -162,6 +162,15 @@ protected override void OnRoomUpdated()
 
             const double fade_time = 50;
 
+            var ruleset = rulesets.GetRuleset(Room.Settings.RulesetID);
+
+            var currentModeRank = User.User?.GetStatisticsFor(ruleset)?.GlobalRank;
+
+            // fallback to current mode rank for testing purposes.
+            currentModeRank ??= User.User?.CurrentModeRank;
```
Comment (peppy) [inline osu.Game/Screens/OnlinePlay/Multiplayer/Participants/ParticipantPanel.cs:170]: why? can't the tests be fixed? leaving this is potentially going to make the incorrect rank for a user show if they have never played in the ruleset of the room they are participating in.

Comment (frenzibyte) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Participants/ParticipantPanel.cs:170]: hmm, yeah I was unsure about it, maybe making `User` expose the statistics-per-ruleset dictionary for tests to add custom ranks to.

File: osu.Game/Users/User.cs [hunk 3]
```diff
@@ -178,6 +184,11 @@ public class KudosuCount
 
         private UserStatistics statistics;
 
+        /// <summary>
+        /// The user statistics of the ruleset specified within the API request.
+        /// If the user is fetched from a <see cref="GetUsersRequest"/> or similar
+        /// (i.e. is a user compact instance), use <see cref="GetStatisticsFor"/> instead.
+        /// </summary>
```
Comment (peppy) [inline osu.Game/Users/User.cs:191]:
```suggestion
        /// User statistics for the requested ruleset (in the case of a <see cref="GetStatisticsFor"/> response).
        /// Null otherwise.
        /// </summary>
```

also mark as `[CanBeNull]` or make the class nullable (whichever is less involved i guess).

File: osu.Game/Users/User.cs [hunk 4]
```diff
@@ -228,13 +239,35 @@ public class UserAchievement
...
+        }
+
+        private UserStatistics parseStatisticsFor(RulesetInfo ruleset)
+        {
+            if (!(otherProperties.TryGetValue($"statistics_{ruleset.ShortName}", out var token)))
+                return new UserStatistics();
+
+            var settings = JsonSerializableExtensions.CreateGlobalSettings();
+            settings.DefaultValueHandling = DefaultValueHandling.Include;
+            return token.ToObject<UserStatistics>(JsonSerializer.Create(settings));
```
Comment (peppy) [inline osu.Game/Users/User.cs:270]: This isn't how we should be doing this. Figure a way to read the json into a `Dictionary<string, Userstatistics>` instead (where the `string` is the `ShortName`). And remove the cache and everything else.

File: osu.Game/Users/User.cs [hunk 5]
```diff
@@ -249,6 +282,14 @@ public class UserHistoryCount
             Id = 0
         };
 
+        public bool Equals(User other)
```
Comment (peppy) [inline osu.Game/Users/User.cs:285]: Why was this moved from the end of the file?

File: osu.Game/Users/UserStatistics.cs [hunk 6]
```diff
@@ -26,6 +27,13 @@ public struct LevelInfo
             public int Progress;
         }
 
+        /// <remarks>
+        /// This must only be used when coming from condensed user responses (e.g. from <see cref="GetUsersRequest"/>), otherwise use <code>Ranks.Global</code>.
+        /// </remarks>
+        // todo: this should likely be moved to a separate UserStatisticsCompact class at some point.
```
Comment (peppy) [inline osu.Game/Users/UserStatistics.cs:33]: this whole comment is ..

have you checked web side how this is done? it sounds like there should be a separate class for the statistics in that new storage area. aka. your todo but not a todo and part of this PR?

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Users/User.cs (hunk 5 above)
Comment (frenzibyte) [reply osu.Game/Users/User.cs:285]: was re-ordering the class a bit, made it to being "properties then user logic then nested classes and enums".

File: osu.Game/Users/UserStatistics.cs (hunk 6 above)
Comment (frenzibyte) [reply osu.Game/Users/UserStatistics.cs:33]: hmm, now that you've mentioned it, making a `UserStatisticsCompact` class would be at least simpler to do than making a `UserCompact`, will add one here.

File: osu.Game/Users/User.cs (hunk 3 above)
Comment (frenzibyte) [reply osu.Game/Users/User.cs:191]: `Statistics` is never null:
https://github.com/ppy/osu/blob/cca1bac67d56d9d261149eb6f497fee824a69811/osu.Game/Users/User.cs#L192-L195

Comment (peppy) [reply osu.Game/Users/User.cs:191]: then make it read "otherwise empty".

but double check this. in other cases we've seen serialisers potentially set fields to null during the deserialisation step.

Comment (frenzibyte) [reply osu.Game/Users/User.cs:191]: it'll regardless return an empty `UserStatistics` for `Statistics` getter, so no need for double checking this unless I misunderstand.

File: osu.Game/Users/User.cs (hunk 4 above)
Comment (frenzibyte) [reply osu.Game/Users/User.cs:270]: Yeah I was struggling on google looking for a way to do this, so I fell back into using a `JsonExtensionData`, will see if a JSON converter would do.

Comment (peppy) [reply osu.Game/Users/User.cs:270]: Are you sure a dictionary doesn't just work? I know we've dealt with a similar structure somewhere else before, so it may pay to start by looking at other responses in this project to find a similar case.

Comment (frenzibyte) [reply osu.Game/Users/User.cs:270]: Well, the problem is that `statistics_{osu/taiko/...}` are direct children of the main `User` json object:
```json
{
    "users": [
        {
            "avatar_url": "https://a.ppy.sh/14210502?1597999788.jpeg",
            "country_code": "SA",
            "default_group": "default",
            "id": 14210502,
            "statistics_fruits": {
                 ...
            },
            "statistics_mania": {
                 ...
            },
            "statistics_osu": {
                 ...
            },
            "statistics_taiko": {
                 ...
            },
        ...
        }
    ]
}
```

So I presume a proper way would be making sometype of custom transformer that deserializes the usual properties to a `User` instance, and then deserialize the statistics and pass them to that instance, which seems to be too much. 

But I've simplified this into parsing the statistics on deserialization (via `OnDeserialized` attribute), which feels pretty good to me than this unecessary caching logic.

---

I would say prehaps osu-web should've included these statistics inside one JSON dictionary there, maybe worth reporting issue there if feasible?

Comment (peppy) [reply osu.Game/Users/User.cs:270]: yep

Comment (frenzibyte) [reply osu.Game/Users/User.cs:270]: Done https://github.com/ppy/osu-web/issues/7197, will hold off from pushing until that's addressed.

**Comment by peppy** [comment]
osu-web side fix is deployed on production and dev.

**Comment by bdach** [comment]
Aside from the review to be addressed, this pull request requires conflict resolution.

**Comment by frenzibyte** [comment]
Yes I'm aware of it, will update it soon enough (was gonna do so yesterday but I got stuck on a bad issue and got the direction forward today).

**Comment by peppy** [comment]
See https://github.com/ppy/osu-web/pull/7209, which should resolve this one osu-web side.

File: osu.Game/Users/UserStatistics.cs [hunk 7]
```diff
@@ -26,17 +26,22 @@ public struct LevelInfo
...
             set => Ranks.Global = value;
         }
 
-        [JsonProperty(@"rank")]
-        public UserRanks Ranks;
+        [JsonProperty(@"pp")]
+        public decimal? PP;
+
+        [JsonProperty(@"pp_rank")]
+        public int PPRank;
```
Comment (frenzibyte) [inline osu.Game/Users/UserStatistics.cs:44]: I've intentionally made this an own property instead of redirecting its value to `Ranks.Global`, as the newly added `global_rank` property serves the same purpose, and it looks to be included in every user statistics API responses as well (`GetUserRequest`, `GetUsersRequest`, `GetUserRankingsRequest`).

I've ensured there were no components affected by this change, but all `Ranks.Global` usages were coming from responses of `GetUserRequest` (where `rank` would be included), i.e. there were no components relying on the `pp_rank` property, so nothing should break there.

File: osu.Game/Users/UserStatistics.cs [hunk 8]
```diff
@@ -26,17 +26,22 @@ public struct LevelInfo
...
 
-        [JsonProperty(@"pp")]
-        public decimal? PP;
+        [JsonProperty(@"rank")]
+        public UserRanks Ranks;
 
-        [JsonProperty(@"pp_rank")] // the API sometimes only returns this value in condensed user responses
-        private int? rank
+        // eventually UserRanks object will be completely replaced with separate global and country rank properties, see https://github.com/ppy/osu-web/blob/cb79bb72186c8f1a25f6a6f5ef315123decb4231/app/Transformers/UserStatisticsTransformer.php#L53.
+        // but for now, always point UserRanks.Global to the global_rank property, as that is included solely for requests like GetUsersRequest.
```
Comment (frenzibyte) [inline osu.Game/Users/UserStatistics.cs:33]: I was initially planning to separate the user statistics class passed to `RulesetsStatistics` from this one, but it appears that `global_rank` is included in every user statistics response, so I decided to just put it here and replace it from the existing `pp_rank` property. (and the separation didn't look that sensible to me either way)

File: osu.Game/Users/UserStatistics.cs (hunk 7 above)
Comment (peppy) [reply osu.Game/Users/UserStatistics.cs:44]: This isn't being used anywhere, is there a reason to keep it around? If we are, it should probably not be called `PPRank` since that's quite nondescript.

File: osu.Game/Users/UserStatistics.cs [hunk 9]
```diff
@@ -26,17 +26,22 @@ public struct LevelInfo
...
+        [JsonProperty(@"global_rank")]
+        private int? globalRank
         {
             set => Ranks.Global = value;
```
Comment (peppy) [inline osu.Game/Users/UserStatistics.cs:37]: I'm not sure I understand how this works. You have `Ranks` above being populated from the incoming request and then overwritten by this? Is some kind of documentation here missing?

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Users/UserStatistics.cs (hunk 9 above)
Comment (frenzibyte) [reply osu.Game/Users/UserStatistics.cs:37]: It's done in a way that if a `global_rank` property exists, then let it overwrite `Ranks.Global`, because, on cases like old saved tournament bracket json files, the global rank would still be stored at the JSON property `rank.global` instead of `global_rank`.

File: osu.Game/Users/UserStatistics.cs (hunk 7 above)
Comment (frenzibyte) [reply osu.Game/Users/UserStatistics.cs:44]: There are no usages to it and can be removed, just felt like it may be used further on for getting an "absolute global rank based on pp".

Also yeah not sure about the name either, just took it directly from the property itself, maybe `RankInPP` or `AbsoluteRank`.

File: osu.Game/Users/UserStatistics.cs (hunk 9 above)
Comment (peppy) [reply osu.Game/Users/UserStatistics.cs:37]: Is that the only case? Sounds like something that should be handled locally if so?

Comment (frenzibyte) [reply osu.Game/Users/UserStatistics.cs:37]: I think so, not sure how to handle it locally though.

Comment (peppy) [reply osu.Game/Users/UserStatistics.cs:37]: Making it perform the fetch again if null is the easiest method. Assuming that's the only usage.

Comment (frenzibyte) [reply osu.Game/Users/UserStatistics.cs:37]: Alright, I've done it locally as you said. Shall I nuke `Ranks.Global` and let all usages use a `GlobalRank` field instead? and if so, shall I do the same with `Ranks.Country` and remove the entire `UserRanks` object?

I was initially going to do that but felt like it may fill the PR with somewhat irrelevant file changes.

Comment (peppy) [reply osu.Game/Users/UserStatistics.cs:37]: Is this change not pushed?

Comment (frenzibyte) [reply osu.Game/Users/UserStatistics.cs:37]: hmm, due to the issues found in https://github.com/ppy/osu/pull/11799, the user IDs may likely not be set to the correct ones, causing the re-fetching to not work as intended. 

I believe the only way would be to either leave this as-is or make a custom JSON converter or otherwise to migrate `ranks.global` to `global_rank` in tournament client. What do you think?

Comment (peppy) [reply osu.Game/Users/UserStatistics.cs]: The failures you found are likely not an issue in any normal usage of the system. I would ignore it.

File: osu.Game/Users/UserStatistics.cs (hunk 7 above)
Comment (bdach) [reply osu.Game/Users/UserStatistics.cs:41]: What does "absolute global rank based on pp" mean? As opposed to what? When is rank not based on pp?

File: osu.Game/Users/UserStatistics.cs [hunk 10]
```diff
@@ -26,18 +26,20 @@ public struct LevelInfo
...
-        [JsonProperty(@"pp")]
-        public decimal? PP;
-
-        [JsonProperty(@"pp_rank")] // the API sometimes only returns this value in condensed user responses
-        private int? rank
-        {
-            set => Ranks.Global = value;
-        }
+        [JsonProperty(@"global_rank")]
+        public int? GlobalRank;
```
Comment (bdach) [inline osu.Game/Users/UserStatistics.cs:30]: So there are three properties now for ranks? `GlobalRank`, `Ranks` and `PPRank`? Which one is valid when? Why do none of those have xmldocs (or obsoletion notices, if they apply) at all? This model design borders on intentional obfuscation.

File: osu.Game/Screens/OnlinePlay/Multiplayer/Participants/ParticipantPanel.cs [hunk 11]
```diff
@@ -171,11 +177,7 @@ protected override void OnRoomUpdated()
 
             // If the mods are updated at the end of the frame, the flow container will skip a reflow cycle: https://github.com/ppy/osu-framework/issues/4187
             // This looks particularly jarring here, so re-schedule the update to that start of our frame as a fix.
-            Schedule(() =>
-            {
-                var ruleset = rulesets.GetRuleset(Room.Settings.RulesetID).CreateInstance();
-                userModsDisplay.Current.Value = User.Mods.Select(m => m.ToMod(ruleset)).ToList();
-            });
+            Schedule(() => userModsDisplay.Current.Value = User.Mods.Select(m => m.ToMod(ruleset.CreateInstance())).ToList());
```
Comment (bdach) [inline osu.Game/Screens/OnlinePlay/Multiplayer/Participants/ParticipantPanel.cs:180]: The previous code created the ruleset once. This code will create it every time. Please revert. I'd hope reverting won't actually change behaviour.

File: osu.Game.Tournament/TournamentGameBase.cs [hunk 12]
```diff
@@ -211,7 +211,7 @@ private bool addBeatmaps()
             return addedInfo;
         }
 
-        public void PopulateUser(User user, Action success = null, Action failure = null)
+        public void PopulateUser(User user, Action success = null, Action failure = null, bool immediate = false)
```
Comment (bdach) [inline osu.Game.Tournament/TournamentGameBase.cs:214]: These changes are from #11799, which this PR is not marked as dependent on.

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Users/UserStatistics.cs (hunk 7 above)
Comment (frenzibyte) [reply osu.Game/Users/UserStatistics.cs:41]: Ignore `in pp`, it's just that this PP rank *presumably* has the user's absolute global rank, regardless of whether the user plays on the ruleset or not, opposed to `global_rank` which returns `null` if the user *isn't ranked* in said ruleset.

File: osu.Game/Users/UserStatistics.cs (hunk 10 above)
Comment (frenzibyte) [reply osu.Game/Users/UserStatistics.cs:30]: `GlobalRank` is what will always stay there, `Ranks.Global` is what's likely to be nuked in favour of `GlobalRank`, `PPRank` is still in discussion whether to remove it, or keep it and rename it to better descriptive name.

File: osu.Game/Screens/OnlinePlay/Multiplayer/Participants/ParticipantPanel.cs (hunk 11 above)
Comment (frenzibyte) [reply osu.Game/Screens/OnlinePlay/Multiplayer/Participants/ParticipantPanel.cs:180]: fair enough

File: osu.Game/Users/UserStatistics.cs (hunk 10 above)
Comment (frenzibyte) [reply osu.Game/Users/UserStatistics.cs:30]: Actually, `Ranks.Global` is already removed, so you can scratch that second one. Could make `Ranks` private and expose a `CountryRank` field as well, if that's good for you?

Comment (frenzibyte) [reply osu.Game/Users/UserStatistics.cs:30]: I've went ahead and done it, should read better now.

File: osu.Game/Users/User.cs (hunk 5 above)
Comment (peppy) [reply osu.Game/Users/User.cs:259]: We generally put all the system functions right at the bottom (dispose/equality/etc).

File: osu.Game/Users/UserStatistics.cs (hunk 7 above)
Comment (peppy) [reply osu.Game/Users/UserStatistics.cs:46]: I don't get it.

Comment (peppy) [reply osu.Game/Users/UserStatistics.cs:46]: This looks like a left-over on osu-web's end. We'll remove `pp_rank` in the next web release, so it can be removed here as well.

**Review by peppy — APPROVED** [review]
